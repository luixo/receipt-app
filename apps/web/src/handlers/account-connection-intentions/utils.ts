import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
