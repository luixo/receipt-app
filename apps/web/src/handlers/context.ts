import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
