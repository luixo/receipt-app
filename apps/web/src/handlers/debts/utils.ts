import type { Database, DebtsId, SimpleInsertObject } from "~db";
import { nonNullishGuard } from "~utils";

export const upsertAutoAcceptedDebts = async (
	database: Database,
	debts: (SimpleInsertObject<"debts"> & { isNew: boolean })[],
) => {
	const fetchedDebts = await database
		.selectFrom("debts")
		.where((eb) =>
			eb.or(
				debts.map((debt) =>
					!debt.isNew
						? eb.and({
								id: debt.id,
								ownerAccountId: debt.ownerAccountId,
						  })
						: debt.receiptId
						? eb.and({
								ownerAccountId: debt.ownerAccountId,
								userId: debt.userId,
								receiptId: debt.receiptId,
						  })
						: eb("debts.ownerAccountId", "is", null),
				),
			),
		)

		.select([
			"debts.id",
			"debts.ownerAccountId",
			"debts.userId",
			"debts.receiptId",
		])
		.execute();
	const counterpartyDebts = debts.map((debt) =>
		fetchedDebts.find((fetchedDebt) => {
			if (debt.isNew) {
				if (!debt.receiptId || !fetchedDebt.receiptId) {
					return false;
				}
				return (
					fetchedDebt.ownerAccountId === debt.ownerAccountId &&
					fetchedDebt.userId === debt.userId &&
					fetchedDebt.receiptId === debt.receiptId
				);
			}
			return (
				fetchedDebt.ownerAccountId === debt.ownerAccountId &&
				fetchedDebt.id === debt.id
			);
		}),
	);
	const existentDebts = debts
		.map((nextDebt, index) =>
			counterpartyDebts[index]
				? ([nextDebt, counterpartyDebts[index]!] as const)
				: null,
		)
		.filter(nonNullishGuard);
	const nonExistentDebts = debts
		.map((nextDebt, index) => (counterpartyDebts[index] ? null : nextDebt))
		.filter(nonNullishGuard);
	const [newDebts, ...updatedDebts] = await Promise.all([
		nonExistentDebts.length === 0
			? ([] as { id: DebtsId }[])
			: database
					.insertInto("debts")
					.values(nonExistentDebts.map(({ isNew, ...debt }) => debt))
					.returning(["debts.id"])
					.execute(),
		...existentDebts.map(([nextDebt, currentDebt]) =>
			database
				.updateTable("debts")
				.set({
					amount: nextDebt.amount,
					timestamp: nextDebt.timestamp,
					currencyCode: nextDebt.currencyCode,
					receiptId: nextDebt.receiptId,
					lockedTimestamp: nextDebt.lockedTimestamp,
				})
				.where((eb) =>
					eb.and({
						id: currentDebt.id,
						ownerAccountId: currentDebt.ownerAccountId,
						userId: currentDebt.userId,
					}),
				)
				.returning(["debts.id", "debts.receiptId"])
				.executeTakeFirst(),
		),
	]);
	return { newDebts, updatedDebts };
};
