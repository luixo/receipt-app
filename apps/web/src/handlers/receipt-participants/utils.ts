import type { Selection } from "kysely";

import type { ReceiptsId, UsersId } from "~db/models";
import type {
	Database,
	ReceiptsDatabase,
	ReceiptsSelectExpression,
} from "~db/types";

export const getReceiptParticipant = <
	SE extends ReceiptsSelectExpression<"receiptParticipants">,
>(
	database: Database,
	userId: UsersId,
	receiptId: ReceiptsId,
	selectExpression: SE[],
): Promise<
	Selection<ReceiptsDatabase, "receiptParticipants", SE> | undefined
> =>
	database
		.selectFrom("receiptParticipants")
		.where((eb) => eb.and({ receiptId, userId }))
		.select(selectExpression)
		.limit(1)
		.executeTakeFirst();
