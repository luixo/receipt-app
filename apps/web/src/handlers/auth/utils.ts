import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
