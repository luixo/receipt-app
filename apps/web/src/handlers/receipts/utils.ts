import type { QueryCreator } from "kysely";
import type { z } from "zod";

import type { AccountsId } from "~db/models";
import type { Database, ReceiptsDatabase } from "~db/types";
import type { assignableRoleSchema } from "~web/handlers/validation";
import { roleSchema } from "~web/handlers/validation";

export type Role = z.infer<typeof roleSchema>;
export type AssignableRole = z.infer<typeof assignableRoleSchema>;

export const getAccessRole = async (
	database: Database,
	receipt: { ownerAccountId: string; id: string },
	accountId: string,
): Promise<Role | undefined> => {
	if (receipt.ownerAccountId === accountId) {
		return "owner";
	}
	const participant = await database
		.selectFrom("receiptParticipants")
		.innerJoin("users", (jb) =>
			jb.onRef("users.id", "=", "receiptParticipants.userId"),
		)
		.innerJoin("accounts", (jb) =>
			jb.onRef("accounts.id", "=", "users.connectedAccountId"),
		)
		.where((eb) =>
			eb.and({
				"accounts.id": accountId,
				receiptId: receipt.id,
			}),
		)
		.select("receiptParticipants.role")
		.limit(1)
		.executeTakeFirst();
	if (!participant) {
		return;
	}
	const parsed = roleSchema.safeParse(participant.role);
	/* c8 ignore start */
	if (!parsed.success) {
		// TODO: add database-level validation
		return;
	}
	/* c8 ignore stop */
	return parsed.data;
};

export const getOwnReceipts = (
	database: Database | QueryCreator<ReceiptsDatabase>,
	ownerAccountId: AccountsId,
) =>
	database.selectFrom("receipts").where("ownerAccountId", "=", ownerAccountId);

export const getParticipantsReceipts = (
	database: Database | QueryCreator<ReceiptsDatabase>,
	ownerAccountId: AccountsId,
) =>
	database
		.selectFrom("users")
		.where((eb) =>
			eb("users.connectedAccountId", "=", ownerAccountId).and(
				"users.ownerAccountId",
				"<>",
				ownerAccountId,
			),
		)
		.innerJoin("receiptParticipants", (jb) =>
			jb.onRef("receiptParticipants.userId", "=", "users.id"),
		)
		.innerJoin("receipts", (jb) =>
			jb.onRef("receipts.id", "=", "receiptParticipants.receiptId"),
		);
