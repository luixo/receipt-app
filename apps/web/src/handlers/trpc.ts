import { TRPCError, initTRPC } from "@trpc/server";
import { isNonNullish, unique } from "remeda";

import { AUTH_COOKIE } from "~app/utils/auth";
import {
	PRETEND_USER_STORE_NAME,
	pretendUserSchema,
} from "~app/utils/store/pretend-user";
import type { AccountsId } from "~db/models";
import { transformer } from "~utils/transformer";
import {
	SESSION_EXPIRATION_DURATION,
	SESSION_SHOULD_UPDATE_EVERY,
	getExpirationDate,
} from "~web/handlers/auth/utils";
import { queueCallFactory } from "~web/handlers/batch";
import type { NetContext, UnauthorizedContext } from "~web/handlers/context";
import { formatErrorMessage } from "~web/handlers/errors";
import { sessionIdSchema } from "~web/handlers/validation";
import { getCookie, setCookie } from "~web/utils/cookies";
import { getReqHeader } from "~web/utils/headers";

export const t = initTRPC.context<UnauthorizedContext>().create({
	transformer,
	errorFormatter: (opts) => {
		const { shape, error } = opts;
		return { ...shape, message: formatErrorMessage(error, shape.message) };
	},
});

export const unauthProcedure = t.procedure.use(
	async ({ ctx, type, path, next }) => {
		const start = Date.now();
		const result = await next();
		const duration = Date.now() - start;
		const options = {
			path,
			type,
			durationMs: duration,
		};
		if (result.ok) {
			ctx.logger.trace(options, "OK request timing:");
		} else {
			ctx.logger.trace(options, "Non-OK request timing:");
		}

		return result;
	},
);

const getAuthToken = (ctx: UnauthorizedContext) =>
	getCookie(getReqHeader(ctx, "cookie"), AUTH_COOKIE);

const getPretendAccountEmail = (ctx: NetContext): string | undefined => {
	const pretendUserString = getCookie(
		getReqHeader(ctx, "cookie"),
		PRETEND_USER_STORE_NAME,
	);
	if (!pretendUserString) {
		return;
	}
	const user = pretendUserSchema.parse(JSON.parse(pretendUserString));
	return user.email;
};

const queueSession = queueCallFactory<
	UnauthorizedContext,
	{ authToken: string },
	{
		realAuth: {
			accountId: AccountsId;
			email: string;
		};
		auth: {
			accountId: AccountsId;
			email: string;
		};
		role: string | undefined;
	}
>(
	(ctx) => async (inputs) => {
		const authTokens = unique(inputs.map(({ authToken }) => authToken));
		const pretendAccountEmail = getPretendAccountEmail(ctx);
		const [sessions, pretendAccount] = await Promise.all([
			ctx.database
				.selectFrom("sessions")
				.innerJoin("accounts", (qb) =>
					qb.onRef("accounts.id", "=", "sessions.accountId"),
				)
				.select([
					"sessions.accountId",
					"accounts.email",
					"accounts.role",
					"sessions.expirationTimestamp",
					"sessions.sessionId",
				])
				.where((eb) =>
					eb("sessions.sessionId", "in", authTokens).and(
						"sessions.expirationTimestamp",
						">",
						new Date(),
					),
				)
				.execute(),
			pretendAccountEmail
				? ctx.database
						.selectFrom("accounts")
						.where("accounts.email", "=", pretendAccountEmail)
						.select(["accounts.id", "accounts.email"])
						.limit(1)
						.executeTakeFirst()
				: undefined,
		]);
		const sessionsOrErrors = inputs.map((input) => {
			const matchedSession = sessions.find(
				(session) => session.sessionId === input.authToken,
			);
			if (!matchedSession) {
				return new TRPCError({
					code: "UNAUTHORIZED",
					message: "Session id mismatch",
				});
			}
			const auth = {
				accountId: matchedSession.accountId,
				email: matchedSession.email,
			};
			if (
				pretendAccount &&
				matchedSession.role === "admin" &&
				!getReqHeader(ctx, "x-keep-real-auth")
			) {
				return {
					realAuth: auth,
					auth: {
						accountId: pretendAccount.id,
						email: pretendAccount.email,
					},
					role: matchedSession.role as "admin",
				};
			}
			return {
				realAuth: auth,
				auth,
				role: matchedSession.role ?? undefined,
			};
		});
		const updateableSessions = authTokens
			.map((authToken) => {
				const matchedSession = sessions.find(
					(session) => session.sessionId === authToken,
				);
				if (
					!matchedSession ||
					matchedSession.expirationTimestamp.valueOf() - Date.now() >=
						SESSION_EXPIRATION_DURATION - SESSION_SHOULD_UPDATE_EVERY
				) {
					return;
				}
				return {
					authToken,
					nextExpirationTimestamp: getExpirationDate(),
				};
			})
			.filter(isNonNullish);
		void ctx.database.transaction().execute(async (tx) => {
			await Promise.all(
				updateableSessions.map(
					async ({ authToken, nextExpirationTimestamp }) => {
						await tx
							.updateTable("sessions")
							.set({ expirationTimestamp: nextExpirationTimestamp })
							.where("sessionId", "=", authToken)
							.executeTakeFirst();
						setCookie(ctx, AUTH_COOKIE, authToken, {
							expires: nextExpirationTimestamp,
						});
					},
				),
			);
		});
		return sessionsOrErrors;
	},
	{
		getKey: (ctx) =>
			[
				// If we've got here, auth token surely exists
				// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
				getAuthToken(ctx)!,
				getReqHeader(ctx, "x-keep-real-auth") ?? "",
			]
				.filter(Boolean)
				.join("/"),
	},
);

export const authProcedure = unauthProcedure.use(async ({ ctx, next }) => {
	const authToken = getAuthToken(ctx);
	if (typeof authToken !== "string" || !authToken) {
		throw new TRPCError({
			code: "UNAUTHORIZED",
			message: "No token provided",
		});
	}
	const uuidVerification = sessionIdSchema.safeParse(authToken);
	if (!uuidVerification.success) {
		throw new TRPCError({
			code: "UNAUTHORIZED",
			message: "Session id mismatch",
		});
	}
	const { realAuth, auth, role } = await queueSession({
		ctx,
		input: { authToken },
		signal: new AbortController().signal,
	});
	return next({
		ctx: {
			...ctx,
			logger: ctx.logger.child({ accountId: auth.accountId }),
			realAuth,
			auth,
			authToken,
			role,
		},
	});
});

export const adminProcedure = authProcedure.use(async ({ ctx, next }) => {
	if (ctx.role !== "admin") {
		throw new TRPCError({
			code: "UNAUTHORIZED",
			message: "Admin procedure is available only if you're an admin",
		});
	}
	return next({ ctx: { ...ctx, auth: ctx.realAuth } });
});
