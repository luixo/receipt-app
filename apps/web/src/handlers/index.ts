import { t } from "~web/handlers/trpc";

import { router as accountRouter } from "./account/index";
import { router as accountConnectionIntentionsRouter } from "./account-connection-intentions/index";
import { router as accountSettingsRouter } from "./account-settings/index";
import { router as adminRouter } from "./admin/index";
import { router as authRouter } from "./auth/index";
import { router as currencyRouter } from "./currency/index";
import { router as debtIntentionsRouter } from "./debt-intentions/index";
import { router as debtsRouter } from "./debts/index";
import { router as receiptItemConsumersRouter } from "./receipt-item-consumers/index";
import { router as receiptItemPayersRouter } from "./receipt-item-payers/index";
import { router as receiptItemsRouter } from "./receipt-items/index";
import { router as receiptParticipantsRouter } from "./receipt-participants/index";
import { router as receiptsRouter } from "./receipts/index";
import { router as resetPasswordIntentionsRouter } from "./reset-password-intentions/index";
import { router as sessionsRouter } from "./sessions/index";
import { router as usersRouter } from "./users/index";
import { router as utilsRouter } from "./utils/index";

export const router = t.router({
	// No auth
	sessions: sessionsRouter,
	auth: authRouter,
	resetPasswordIntentions: resetPasswordIntentionsRouter,
	utils: utilsRouter,
	// Auth
	account: accountRouter,
	accountSettings: accountSettingsRouter,
	receipts: receiptsRouter,
	receiptItems: receiptItemsRouter,
	users: usersRouter,
	receiptParticipants: receiptParticipantsRouter,
	currency: currencyRouter,
	receiptItemConsumers: receiptItemConsumersRouter,
	receiptItemPayers: receiptItemPayersRouter,
	accountConnectionIntentions: accountConnectionIntentionsRouter,
	debts: debtsRouter,
	debtIntentions: debtIntentionsRouter,
	// Admin
	admin: adminRouter,
});
