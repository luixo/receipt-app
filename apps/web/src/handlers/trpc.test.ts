import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { PRETEND_USER_STORE_NAME } from "~app/utils/store/pretend-user";
import { createAuthContext, createContext } from "~tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
} from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import {
	SESSION_EXPIRATION_DURATION,
	SESSION_SHOULD_UPDATE_EVERY,
} from "~web/handlers/auth/utils";
import { t } from "~web/handlers/trpc";
import { getHeadersEntries } from "~web/utils/headers";

import { router } from "./index";

const createCaller = t.createCallerFactory(router);

describe("procedures", () => {
	describe("unauth procedure", () => {
		test("success logged data", async ({ ctx }) => {
			ctx.logger.level = "trace";

			const caller = createCaller(createContext(ctx));
			await caller.sessions.cleanup();
			const loggedMessages = ctx.logger.getMessages();
			expect(Array.isArray(loggedMessages)).toBe(true);
			const loggedProcedureMessage = loggedMessages[
				loggedMessages.length - 1
			] as [
				{
					durationMs: number;
					path: string;
					type: string;
				},
				string,
			];
			expect(loggedProcedureMessage[0].durationMs).toBeTypeOf("number");
			expect(loggedProcedureMessage).toStrictEqual<
				typeof loggedProcedureMessage
			>([
				{
					durationMs: loggedProcedureMessage[0].durationMs,
					path: "sessions.cleanup",
					type: "mutation",
				},
				"OK request timing:",
			]);
		});
		test("failed logged data", async ({ ctx }) => {
			ctx.logger.level = "trace";

			const caller = createCaller(createContext(ctx));
			await caller.account.get().catch((e) => e);
			const loggedMessages = ctx.logger.getMessages();
			expect(Array.isArray(loggedMessages)).toBe(true);
			const loggedProcedureMessage = loggedMessages[
				loggedMessages.length - 1
			] as [
				{
					durationMs: number;
					path: string;
					type: string;
				},
				string,
			];
			expect(loggedProcedureMessage[0].durationMs).toBeTypeOf("number");
			expect(loggedProcedureMessage).toStrictEqual<
				typeof loggedProcedureMessage
			>([
				{
					durationMs: loggedProcedureMessage[0].durationMs,
					path: "account.get",
					type: "query",
				},
				"Non-OK request timing:",
			]);
		});
	});

	describe("auth procedure", () => {
		test("no token provided", async ({ ctx }) => {
			const caller = createCaller(createContext(ctx));
			await expectTRPCError(
				() => caller.account.get(),
				"UNAUTHORIZED",
				"No token provided",
			);
		});

		test("invalid uuid", async ({ ctx }) => {
			const caller = createCaller(
				createContext(ctx, {
					headers: { cookie: "authToken=fake" },
				}),
			);
			await expectTRPCError(
				() => caller.account.get(),
				"UNAUTHORIZED",
				"Session id mismatch",
			);
		});

		test("non-existent session", async ({ ctx }) => {
			// Verify that other accounts don't affect the result
			await insertAccountWithSession(ctx);

			const caller = createCaller(
				createContext(ctx, {
					headers: { cookie: `authToken=${faker.string.uuid()}` },
				}),
			);
			await expectTRPCError(
				() => caller.account.get(),
				"UNAUTHORIZED",
				"Session id mismatch",
			);
		});

		test("session is not auto-updated", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx, {
				session: {
					expirationTimestamp: new Date(
						Date.now() +
							(SESSION_EXPIRATION_DURATION - SESSION_SHOULD_UPDATE_EVERY) +
							1000,
					),
				},
			});

			const context = createAuthContext(ctx, sessionId);
			const caller = createCaller(context);
			await expectDatabaseDiffSnapshot(ctx, () => caller.account.get());
			const responseHeaders = getHeadersEntries(context.res.headers);
			expect(responseHeaders).toStrictEqual<typeof responseHeaders>([]);
		});

		test("session is auto-updated", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx, {
				session: {
					expirationTimestamp: new Date(
						Date.now() +
							(SESSION_EXPIRATION_DURATION - SESSION_SHOULD_UPDATE_EVERY) -
							1000,
					),
				},
			});

			const context = createAuthContext(ctx, sessionId);
			const caller = createCaller(context);
			await expectDatabaseDiffSnapshot(ctx, () => caller.account.get());
			const responseHeaders = getHeadersEntries(context.res.headers);
			expect(responseHeaders).toStrictEqual<typeof responseHeaders>([
				[
					"set-cookie",
					`authToken=${sessionId}; Path=/; Expires=Fri, 31 Jan 2020 00:00:00 GMT; HttpOnly; SameSite=Strict`,
				],
			]);
		});

		describe("pretend user", () => {
			describe("context is regular", () => {
				test("with non-admin role", async ({ ctx }) => {
					const { sessionId, accountId } = await insertAccountWithSession(ctx);
					const caller = createCaller(createAuthContext(ctx, sessionId));
					const { account } = await caller.account.get();
					expect(account.id).toEqual(accountId);
				});

				test("with email not found", async ({ ctx }) => {
					const { sessionId, accountId } = await insertAccountWithSession(ctx, {
						account: { role: "admin" },
					});
					const caller = createCaller(
						createAuthContext(ctx, sessionId, {
							headers: {
								cookie: `${PRETEND_USER_STORE_NAME}=${JSON.stringify({
									email: "not@found.com",
								})}`,
							},
						}),
					);
					const { account } = await caller.account.get();
					expect(account.id).toEqual(accountId);
				});

				test("with magic header", async ({ ctx }) => {
					const { sessionId, accountId } = await insertAccountWithSession(ctx, {
						account: { role: "admin" },
					});
					const foreignAccount = await insertAccount(ctx);
					const caller = createCaller(
						createAuthContext(ctx, sessionId, {
							headers: {
								cookie: `${PRETEND_USER_STORE_NAME}=${JSON.stringify({
									email: foreignAccount.email,
								})}`,
								"x-keep-real-auth": "true",
							},
						}),
					);
					const { account } = await caller.account.get();
					expect(account.id).toEqual(accountId);
				});

				test("with malformed cookie", async ({ ctx }) => {
					const { sessionId, accountId } = await insertAccountWithSession(ctx, {
						account: { role: "admin" },
					});
					const foreignAccount = await insertAccount(ctx);
					const caller = createCaller(
						createAuthContext(ctx, sessionId, {
							headers: {
								cookie: `${PRETEND_USER_STORE_NAME}=${JSON.stringify({
									email2: foreignAccount.email,
								})}`,
							},
						}),
					);
					const { account } = await caller.account.get();
					expect(account.id).toEqual(accountId);
				});
			});

			test("context is swapped", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx, {
					account: { role: "admin" },
				});
				const foreignAccount = await insertAccount(ctx);
				const caller = createCaller(
					createAuthContext(ctx, sessionId, {
						headers: {
							cookie: `${PRETEND_USER_STORE_NAME}=${JSON.stringify({
								email: foreignAccount.email,
							})}`,
						},
					}),
				);
				const { account } = await caller.account.get();
				expect(account.id).toEqual(foreignAccount.id);
			});
		});
	});

	describe("admin procedure", () => {
		test("user is not admin", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx, {
				account: { role: "foo" },
			});
			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() => caller.admin.accounts(),
				"UNAUTHORIZED",
				"Admin procedure is available only if you're an admin",
			);
		});

		test("user is admin", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx, {
				account: { role: "admin" },
			});
			const caller = createCaller(createAuthContext(ctx, sessionId));
			const accounts = await caller.admin.accounts();
			expect(accounts.length).toBe(0);
		});
	});
});
