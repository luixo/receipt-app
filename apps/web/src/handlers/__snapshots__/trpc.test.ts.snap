// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`procedures > auth procedure > session is auto-updated 1`] = `
{
  "updated": {
    "sessions": {
      "sessionId:47976683-7307-4041-9a33-4f42da683bd8": {
        "expirationTimestamp": 2020-01-31T00:00:00.000Z,
      },
    },
  },
}
`;

exports[`procedures > auth procedure > session is not auto-updated 1`] = `{}`;
